### PR TITLE
[GraphBolt] enable load/save for CSCSamplingGraph

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: graphbolt/include/utils.h [hunk 1]
```diff
@@ -0,0 +1,151 @@
+/**
+ *  Copyright (c) 2023 by Contributors
+ * @file graphbolt/include/csc_sampling_graph.h
```
Comment (jermainewang) [inline graphbolt/include/utils.h:3]: Filename did not match. `utils.h` is perhaps to generic. How about `serialize.h`?

File: graphbolt/include/utils.h [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+template <typename DataT>
+void write_to_archive(
+    torch::serialize::OutputArchive& archive,
+    const std::string& key,
+    const DataT& data) {
+  archive.write(key, data);
+}
+
+/**
+ * @brief Specialization utility function to save vector.
```
Comment (jermainewang) [inline graphbolt/include/utils.h:30]: ... save string vector.

File: graphbolt/include/utils.h [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    torch::serialize::InputArchive& archive,
+    const std::string& key,
+    torch::Tensor& data) {
+  torch::IValue iv_data;
+  archive.read(key, iv_data);
+  data = iv_data.toTensor();
+}
+
+/**
+ * @brief Specialization utility function to read to vector.
```
Comment (jermainewang) [inline graphbolt/include/utils.h:129]: read to string vector

**Review by jermainewang — APPROVED** [review]
LGTM. Do you have a follow-up item to add unittest?

File: graphbolt/include/utils.h [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+/**
+ * @brief Specialization utility function to save vector.
+ * @param archive Output archive.
+ * @param key Key name used in saving.
+ * @param data Vector of string.
+ *
+ *
+*/
+template<>
```
Comment (frozenbugs) [inline graphbolt/include/utils.h:37]:
```suggestion
template <>
```

File: graphbolt/include/utils.h [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  archive.write(key, data);
+}
+
+/**
+ * @brief Specialization utility function to save vector.
+ * @param archive Output archive.
+ * @param key Key name used in saving.
+ * @param data Vector of string.
+ *
+ *
```
Comment (frozenbugs) [inline graphbolt/include/utils.h:35]: remove empty.

File: graphbolt/include/utils.h [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+ * @param archive Output archive.
+ * @param key Key name used in saving.
+ * @param data Data that could be constructed as `torch::IValue`.
+*/
+template <typename DataT>
+void write_to_archive(
+    torch::serialize::OutputArchive& archive,
+    const std::string& key,
+    const DataT& data) {
+  archive.write(key, data);
```
Comment (frozenbugs) [inline graphbolt/include/utils.h:26]: Should we put all implementations in .cc file?

File: graphbolt/include/utils.h [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  data = iv_data.toTensor();
+}
+
+/**
+ * @brief Specialization utility function to read to vector.
+ * @param archive Output archive.
+ * @param key Key name used in saving.
+ * @param data Vector of string.
+*/
+template<>
```
Comment (frozenbugs) [inline graphbolt/include/utils.h:134]:
```suggestion
template <>
```

File: graphbolt/include/csc_sampling_graph.h [hunk 8]
```diff
@@ -66,6 +69,25 @@ struct HeteroInfo {
    * edge_types. The length of it is equal to the number of edges.
    */
   torch::Tensor type_per_edge;
+
+  /**
+  * @brief Magic number to indicate Hetero info version in serialize/
```
Comment (frozenbugs) [inline graphbolt/include/csc_sampling_graph.h:74]: run lintrunner -a

File: graphbolt/include/csc_sampling_graph.h [hunk 9]
```diff
@@ -66,6 +69,25 @@ struct HeteroInfo {
...
+  * deserialize stages.
+  */
+  static constexpr int64_t kHeteroInfo_Serialize_Magic
+    = 0xDD2E60F0F6B4A129;
+
+  /**
+  * @brief Load hetero info from stream.
+  * @param archive Input stream for deserializing.
+  */
```
[the remark this answers is not in the document]
Comment (frozenbugs) [reply graphbolt/include/csc_sampling_graph.h:83]: Should be fixed by auto reformat.

File: graphbolt/include/csc_sampling_graph.h [hunk 10]
```diff
@@ -148,6 +170,25 @@ class CSCSamplingGraph : public torch::CustomClassHolder {
     return hetero_info_->type_per_edge;
   }
 
+  /**
+  * @brief Magic number to indicate graph version in serialize/deserialize
+  * stage.
+  */
+  static constexpr int64_t kCSCSamplingGraph_Serialize_Magic
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply graphbolt/include/csc_sampling_graph.h:177]: just to be clear that each class has a dedicated version number.

File: graphbolt/include/utils.h (hunk 6 above)
Comment (Rhett-Ying) [reply graphbolt/include/utils.h:26]: It's template functions,  I think it's ok to be header only.

**Comment by Rhett-Ying** [comment]
> LGTM. Do you have a follow-up item to add unittest?

I plan to add python-level tests only which will be added in another python PR.

File: graphbolt/include/utils.h (hunk 6 above)
Comment (frozenbugs) [reply graphbolt/include/utils.h]: But there is also template instantiation e.g. `void write_to_archive<std::vector<std::string>>(`.

If you have to put it in header.

Add
`
#ifndef GRAPHBOLT_INCLUDE_UTILS_H_
#define GRAPHBOLT_INCLUDE_UTILS_H_
...
#endif
`

example: https://source.chromium.org/chromium/chromium/src/+/main:base/check_deref.h;drc=cda36986fdb5e85aeeaa6e106e77fea17574ad4a;l=12

File: graphbolt/src/csc_sampling_graph.cc [hunk 11]
```diff
@@ -6,9 +6,32 @@
...
 
+#include "serialize.h"
+
 namespace graphbolt {
 namespace sampling {
 
+void HeteroInfo::Load(torch::serialize::InputArchive& archive) {
+  int64_t magic_num = 0x0;
+  utils::read_from_archive(archive, "HeteroInfo/magic_num", magic_num);
+  TORCH_CHECK(magic_num, kHeteroInfoSerializeMagic);
```
Comment (czkkkkkk) [inline graphbolt/src/csc_sampling_graph.cc:17]:
```suggestion
  TORCH_CHECK(magic_num, kHeteroInfoSerializeMagic) << "Magic number mismatch when loading CSCSamplingGraph";
```

File: graphbolt/src/csc_sampling_graph.cc [hunk 12]
```diff
@@ -48,5 +71,52 @@ c10::intrusive_ptr<CSCSamplingGraph> CSCSamplingGraph::FromCSCWithHeteroInfo(
       num_nodes, indptr, indices, hetero_info);
 }
 
+void CSCSamplingGraph::Load(torch::serialize::InputArchive& archive) {
+  int64_t magic_num = 0x0;
+  utils::read_from_archive(archive, "CSCSamplingGraph/magic_num", magic_num);
+  TORCH_CHECK(magic_num, kCSCSamplingGraphSerializeMagic);
```
Comment (czkkkkkk) [inline graphbolt/src/csc_sampling_graph.cc:77]:
```suggestion
  TORCH_CHECK(magic_num, kCSCSamplingGraphSerializeMagic) << "Magic number mismatch when loading CSCSamplingGraph";
```

**Review by czkkkkkk — COMMENTED** [review]
Is it possible to add unittests in Python?

**Comment by Rhett-Ying** [comment]
> Is it possible to add unittests in Python?

I plan to add test when python-level CSCSamplingGraph is added which also includes python bindings. Or these 2 PRs will have much in common.

**Review by czkkkkkk — APPROVED** [review]

**Pull request merged**